### ISSUE TITLE
Clarify when the `nullable::` annotation can be used and how it interacts with constraints

### ISSUE THREAD

**Issue by popematt** [amzn/ion-schema#38]
The Ion Schema [Grammar](https://amzn.github.io/ion-schema/docs/spec.html#grammar) says that `nullable::` can be used on any type reference. However, the section on [Core Types](https://amzn.github.io/ion-schema/docs/spec.html#core-types) only mentions that `nullable::` can be added to core types, so it's not very clear how `nullable::` is supposed to interact with user types. Here is a list of examples that I have compiled that are unclear or possibly even undefined by the spec.

| Type | Comments |
|:---|:----
| `{ type: nullable::int, not: { type: int } }` | Only allows the values `null`, `null.null`, and `null.int`?
| `{ type: nullable::int, not: nullable:{ type: int } }` | Only allows the values `null`, `null.null`?
| `{ not: nullable::int }` | Includes any non-null, non-int value.
| `{ type: nullable::{ not: int } }` | Includes anything (even nulls) except ints. I think `null.int` is allowed—should it be?
| `{ type: nullable::{ not: nullable::int } }` | Includes anything (even nulls) except for nullable ints. Are `null` and `null.null` allowed by this?
| `{ type: nullable::$int }` | Is this valid?
| `{ type: nullable::int, valid_values:range::[0, max] }` | The `valid_values` constraint automatically requires the value to be not null, so does `nullable` have any effect here?
| `{ type: nullable::{ type: int, valid_values:range::[0, max] } }` | What should nullable mean here? Does the `valid_values` constraint come into play and require that it be not null?
| `{ type: nullable::int, not: { valid_values:[0] } }` | At face value, this looks like it should be a nullable, non-zero int. Is it?
| `{ all_of: [nullable::int, {not:{valid_values:[0] }]}` | Is this equivalent to the prior example?
| `{ type: nullable::{ fields:{ foo:any } } }` | Does this allow any null or does it allow only `null`, `null.null`, and `null.struct`?
|`{ any_of: [nullable::int, nullable::string] }`<br/>`{ type: nullable::{ any_of: [int, string] }`| Should these be equivalent? What values would they actually allow?
|`{ one_of: [nullable::int, nullable::string] }` | Does this exclude untyped nulls? 
|`{ type: nullable::nothing }` | Does this allow any strongly-typed or untyped nulls?

I think that we either need to be more specific about how `nullable::` is supposed to work, or we should we should decide to discourage `nullable::` in favor of making it explicit in the schema when and how a type can be null and non null (ie. make it easy to express `foo` _or_ `null`).

**Comment by popematt** [comment]
### Proposal

In Ion Schema 2.0, the `nullable` annotation shall be removed and `$null_or` will be added. The `$null_or` annotation modifies the _use_ of a type reference to also allow the `null` (or `null.null`) value.

The `$null_or` annotation shall not be applicable on top-level type definitions. It is only applicable to the _use_ of a type, and a top-level type definition is a _declaration_ of a type.

#### Syntax

```
<TYPE_REFERENCE> ::=           <TYPE_NAME>
                   | $null_or::<TYPE_NAME>
                   |           <TYPE_ALIAS>
                   | $null_or::<TYPE_ALIAS>
                   |           <UNNAMED_TYPE_DEFINITION>
                   | $null_or::<UNNAMED_TYPE_DEFINITION>
                   |           <IMPORT_TYPE>
                   | $null_or::<IMPORT_TYPE>
```

#### Evaluation

When the `$null_or` annotation is present on any type reference, it SHALL be evaluated equivalently to the union of `$null` and the annotated type. For example:
```
// This $null_or annotatated type

$null_or::{ type: string, codepoint_length: range::[min, 8] }

// Shall be evaluated as

{
  any_of:[
    $null,
    { type: string, codepoint_length: range::[min, 8] }
  ]
}
```

**Comment by popematt** [comment]
Resolved by #76 in Ion Schema 2.0
